### PR TITLE
fix: [CDS-40944]: prevent YAML to Visual switch on duplicate identifiers in pipelineTemplates

### PR DESCRIPTION
**Pull request by PratyushGar**
#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 6 Issues 
* Code issue 6 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/e864badd0b95ab81c060b15b85f15701/pull-request/10505)

**Review by vkbansal-harness — APPROVED** [review]

File: src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvas.tsx [hunk 1]
```diff
@@ -72,15 +72,14 @@ import GitRemoteDetails from '@common/components/GitRemoteDetails/GitRemoteDetai
...
-import { hasDuplicateIdentifier } from '@common/utils/utils'
 import { usePipelineContext } from '../PipelineContext/PipelineContext'
 import CreatePipelines from '../CreateModal/PipelineCreate'
 import { DefaultNewPipelineId, DrawerTypes } from '../PipelineContext/PipelineActions'
 import PipelineYamlView from '../PipelineYamlView/PipelineYamlView'
 import { RightBar } from '../RightBar/RightBar'
 import StudioGitPopover from '../StudioGitPopover'
 import usePipelineErrors from './PipelineErrors/usePipelineErrors'
-import { getStageIdDetailsMapping } from './PipelineCanvasUtils'
+import { getDuplicateStepsIdentifiersList } from './PipelineCanvasUtils'
```
Comment (udayvunnam) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvas.tsx:82]: nit: several plurals-  steps, identifiers, list
getDuplicateStepIdentifiers or getDuplicateStepIdentifierList

File: src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvas.tsx [hunk 2]
```diff
@@ -739,7 +731,7 @@ export function PipelineCanvas({
         )
       return !(duplicateStepIdentifiersList.length > 0)
     }
-    return false
+    return true
```
Comment (udayvunnam) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvas.tsx:734]: Is this expected? Can this change include a test?

Comment (yc147) [reply src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvas.tsx:734]: This doesn't look write. How can this function return `true` when there are no stages. Function name should be `doesNoHaveDuplicateIdentifiersInYAML` or something like that.

Comment (PratyushGar) [reply src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvas.tsx]: Addressed

**Review by vkbansal-harness — APPROVED** [review]

File: src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvas.tsx [hunk 3]
```diff
@@ -722,24 +721,18 @@ export function PipelineCanvas({
...
-      const duplicateStepIdentifiersList = [] as string[]
-      Object.values(stepsIdMap).forEach(stepIdMap => {
-        const duplicateSteps = hasDuplicateIdentifier(stepIdMap.steps)
-        const duplicateRollbackSteps = hasDuplicateIdentifier(stepIdMap.rollbackSteps)
-        duplicateSteps.value && duplicateStepIdentifiersList.push(...duplicateSteps.duplicateIds)
-        duplicateRollbackSteps.value && duplicateStepIdentifiersList.push(...duplicateRollbackSteps.duplicateIds)
-      })
+      const duplicateStepIdentifiersList = getDuplicateStepIdentifierList(stagesData)
       duplicateStepIdentifiersList.length &&
         showError(
```
Comment (yc147) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvas.tsx:726]: Not the right place to show the error. This function is supposed to run a boolean. Moved the logic to show error out side this function.

**Review by yc147 — CHANGES_REQUESTED** [review]
Not the right place to show the error. This function is supposed to run a boolean. Moved the logic to show error out side this function.

File: src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvasUtils.ts [hunk 4]
```diff
@@ -111,6 +111,10 @@ export const getCategoryItems = (_stageType: string, _selectedStage: StageElemen
   //   }
 }
 
+export const getDuplicateIdentifiers = (idMap: string[]): string[] => {
+  return idMap.filter((element, position, arr) => arr.indexOf(element) !== position)
```
Comment (yc147) [inline src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvasUtils.ts:115]: This algo has non linear time complexity

Comment (yc147) [reply src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvasUtils.ts:115]: Let's keep it for now as the list is not going to be too large. Ideally this should be solved in linear time complexity.

Comment (PratyushGar) [reply src/modules/70-pipeline/components/PipelineStudio/PipelineCanvas/PipelineCanvasUtils.ts:115]: Sure, will address this in the next PR.

**Review by yc147 — APPROVED** [review]

**Review by vkbansal-harness — APPROVED** [review]

**Pull request merged**